### ISSUE TITLE
repeatWhenEmpty operator doesn't work when used in an RSocket

### ISSUE THREAD

**Issue by codependent** [rsocket/rsocket-java#860]
## Expected Behavior
`repeatWhenEmpty()` shoud work as expected

## Actual Behavior
When using that operator subscriptions get stuck in an infinite loop. The same Publisher executes correctly when invoked from an HTTP SpringWebflux endpoint

## Steps to Reproduce
Sample to reproduce the problem: https://github.com/codependent/rsocket-rating-service

Detailed context available on StackOverflow: https://stackoverflow.com/questions/62123120/repeatwhenempty-operator-doesnt-work-in-the-context-of-an-rsocket-websocket-req

* RSocket version(s) used: 1.0.0
* Other relevant libraries versions (eg. `netty`, ...): Spring Boot 2.3.0
* Platform (eg. JVM version (`javar -version`) or Node version (`node --version`)): JDK 11
* OS and version (eg `uname -a`): Darwin MacBook-Pro.local 19.4.0 Darwin Kernel Version 19.4.0

**Comment by OlegDokuka** [comment]
Hey @codependent!

Looking into your example I have a few questions:

1. How is this related to a vanilla RSocket-Java?
2. Have you tested your code in isolation so you confirmed it is not Reactor issue?
3. Have checked the same without using repeateWhenEmpty operator?

Cheers,
Oleh

**Comment by codependent** [comment]
Hi @OlegDokuka,

Well I’m kind of confused here. The code works fine in isolation (main method subscribing the Mono chain, plain http request answered by a GetMapping controller...). The only case in which it doesn’t work is when it is invoked in a Spring Boot GetMapping

**Comment by OlegDokuka** [comment]
Did you try to write the same using vanilla RSocket-Java without Spring Framework? 

Assume I'm trying to reduce the surface of the investigation

Cheers,
Oleh

**Comment by codependent** [comment]
Not really... I’ll do it and come back here with the results. Thanks

**Comment by codependent** [comment]
It seems to happen in a plain RSocket Java application: https://github.com/codependent/rsocket-samples/blob/repeatWhenEmpty/src/main/kotlin/com/codependent/rsocket/RequestResponseRSocket.kt

```
val logger = LoggerFactory.getLogger("com.codependent")

fun main() {

    RSocketServer.create()
        .payloadDecoder(PayloadDecoder.DEFAULT)
        .acceptor { setup, sendingSocket ->
            Mono.just(
                object : AbstractRSocket() {
                    override fun requestResponse(payload: Payload): Mono<Payload> {
                        println("Received ${payload.dataUtf8}")
                        //return Mono.just(DefaultPayload.create("Pong"))
                        return generateRating()
                            .flatMap { Mono.just(DefaultPayload.create(it))}
                    }
                })
        }
        .bind(TcpServerTransport.create(7878))
        .block()!!
        .onClose()

    val client = RSocketConnector.connectWith(TcpClientTransport.create(7878))
        .block()!!

    IntStream.range(0, 10).forEach {
        client.requestResponse(DefaultPayload.create("Ping"))
            .map { it.dataUtf8 }
            .onErrorReturn("error")
            .doOnNext(System.out::println)
            .block()
    }
}


fun generateRating(): Mono<String> {

    return Mono.just("hi there").log()
        .doOnNext {
            logger.info("Next1 {}", it)
        }
        .doOnCancel {
            logger.info("Cancel1")
        }
        .doOnSuccess {
            logger.info("Success1 {}", it)
        }
        .doOnError { throwable ->
            logger.error("Error1 {}", throwable)
        }
        .doOnTerminate {
            logger.info("Terminate1")
        }
        .repeatWhenEmpty(Repeat.times<Any>(1L)).log()
        .doOnNext {
            logger.info("Next2 {}", it)
        }
        .doOnCancel {
            logger.info("Cancel2")
        }
        .doOnSuccess {
            logger.info("Success2 {}", it)
        }
        .doOnError { throwable ->
            logger.error("Error2 {}", throwable)
        }
        .doOnTerminate {
            logger.info("Terminate2")
        }

}
```

I you just comment out `.repeatWhenEmpty(Repeat.times<Any>(1L)).log()` it works. Otherwise it gets stuck in the loop from Operators:

```
	public static void onDiscardMultiple(@Nullable Iterator<?> multiple, boolean knownToBeFinite, Context context) {
		if (multiple == null) return;
		if (!knownToBeFinite) return;

		Consumer<Object> hook = context.getOrDefault(Hooks.KEY_ON_DISCARD, null);
		if (hook != null) {
			try {
				multiple.forEachRemaining(o -> {
```

**Comment by OlegDokuka** [comment]
After all investigation, I'm stating that this is reactors bug which hangs when it comes to discarding 

```

 iterations = Flux.fromStream(LongStream.range(0, Long.MAX_VALUE).boxed());
```

The issue can be reproduced on the following example in isolation:

```
public static void main(String... args) {

		generateRating().block();
}

static Mono<String> generateRating() {

		return Mono.just("hi there")
		           .doOnCancel(() -> logger.info("BeforeRepeatWhen {}()", SignalType.CANCEL))
		           .doOnRequest(r -> logger.info("BeforeRepeatWhen {}({})", SignalType.REQUEST, r))
		           .doOnEach(st -> logger.info("BeforeRepeatWhen {}", st))
		           .repeatWhenEmpty(flux -> flux
				           .doOnCancel(() -> logger.info("InnerRepeatWhen {}()", SignalType.CANCEL))
				           .doOnRequest(r -> logger.info("InnerRepeatWhen {}({})", SignalType.REQUEST, r))
				           .doOnEach(st -> logger.info("InnerRepeatWhen {}", st))
				           .zipWith(Flux.just(1))
		           )
		           .doOnCancel(() -> logger.info("AfterRepeatWhen {}()", SignalType.CANCEL))
		           .doOnRequest(r -> logger.info("AfterRepeatWhen {}({})", SignalType.REQUEST, r))
		           .doOnEach(st -> logger.info("AfterRepeatWhen {}", st))
				.doOnDiscard(Object.class, o -> {
					if (o != null) {
						o.toString();
					}
				});
}
```

This works perfectly in Spring WebFlux since it does no add `onDiscard` handler internally. In contrast in RSocket we wanna make sure there are no leaks since the user operates with a memory sensitive value such as Payload.

A workaround as of now can be the usage of the following override of the `repeatWhenEmpty` method 

```
public final Mono<T> repeatWhenEmpty(int maxRepeat, Function<Flux<Long>, ? extends Publisher<?>> repeatFactory) {
```

which is explicit limits the number of repetitions to a specific number.

Otherwise, it Reactor will try to discard `LongStream.range(0, Long.MAX_VALUE).boxed()` of elements

cc @simonbasle

**Comment by simonbasle** [comment]
I just saw this issue pop up in my feed due to being closed. FYI the root issue was https://github.com/reactor/reactor-core/issues/2196 and it was fixed back in 3.2.19.RELEASE and 3.3.8.RELEASE 👍

**Comment by OlegDokuka** [comment]
yeah! I have seen it! That is why closed this
